### PR TITLE
[AJ-1310] Identify protected Azure workspaces

### PR DESCRIPTION
**Pull request by nawatts**
https://broadworkbench.atlassian.net/browse/AJ-1310

Currently, `isProtectedWorkspace` returns false for all Azure workspaces.
https://github.com/DataBiosphere/terra-ui/blob/de58681e8d432c0193b728b611a8d9dcdb46747f/src/import-data/protected-data-utils.ts#L74-L79

This means that the UI will not allow selecting an Azure workspace to import protected data into.

This updates the UI to identify protected Azure workspaces based on their policies.

### CODE REVIEW

File: src/import-data/ImportDataDestination.ts [hunk 1]
```diff
@@ -111,7 +111,23 @@ export const ImportDataDestination = (props: ImportDataDestinationProps): ReactN
   const immediateImportTypes: ImportRequest['type'][] = ['tdr-snapshot-reference', 'catalog-snapshots'];
   const importMayTakeTime = !immediateImportTypes.includes(importRequest.type);
 
-  const { workspaces, refresh: refreshWorkspaces, loading: loadingWorkspaces } = useWorkspaces();
+  const {
+    workspaces,
+    refresh: refreshWorkspaces,
+    loading: loadingWorkspaces,
+  } = useWorkspaces([
+    'workspace.workspaceId',
```
Comment (jladieu) [inline src/import-data/ImportDataDestination.ts:119]: Just want to double check here, looks like your fields list is drifting considerably from the default list provided in `useWorkspaces` -- if I were doing this I'd be fearful of surprises in later logic from dropping params like `workspace.state`, `workspace.attributes`, etc.

I'm wondering if `useWorkspaces` should expose a `defaultWorkspaceFields` and then callers like this one can start with that and append stuff, eg: `[...defaultWorkspaceFields, 'workspace.workspaceId', <arg3>, <arg4>]`?

That said, I do appreciate _only asking for what we need_ so if my hesitation is uncalled for, by all means, go for it!

**Review by jladieu — APPROVED** [review]

File: src/import-data/ImportDataDestination.ts [hunk 2]
```diff
@@ -111,7 +111,23 @@ export const ImportDataDestination = (props: ImportDataDestinationProps): ReactN
...
+    'workspace.name',
+    // The decision on whether or data can be imported into a workspace is based on the user's level of access
+    // to the workspace and the workspace's authorization domain, protected status and cloud platform.
+    // That information needs to be fetched here.
+    'accessLevel',
+    'policies',
+    'workspace.authorizationDomain',
+    'workspace.bucketName',
+    'workspace.cloudPlatform',
+  ]);
```
Comment (davidangb) [inline src/import-data/ImportDataDestination.ts:130]: nice - this is a big win for performance too!

File: src/import-data/protected-data-utils.test.ts [hunk 3]
```diff
@@ -80,7 +81,22 @@ describe('isProtectedWorkspace', () => {
...
+        {
+          additionalData: [],
+          namespace: 'terra',
+          name: 'protected-data',
+        },
+      ],
+    };
+
+    expect(isProtectedWorkspace(protectedAzureWorkspace)).toBe(true);
+  });
```
Comment (davidangb) [inline src/import-data/protected-data-utils.test.ts:97]: could you also add the negative test, where policies is empty, or policies contains something that is not `protected-data`, and expect `isProtectedWorkspace` to be false?

File: src/import-data/protected-data-utils.ts [hunk 4]
```diff
@@ -68,12 +68,28 @@ export const isProtectedSource = (importRequest: ImportRequest): boolean => {
...
+  const cloudProvider = getCloudProviderFromWorkspace(workspace);
+  switch (cloudProvider) {
+    case 'AZURE':
+      // The WorkspaceWrapper type specifies policies as optional, but policies are loaded by the ImportData component.
+      if (!workspace.policies) {
+        return false;
+      }
+      return workspace.policies.some((policy) => policy.namespace === 'terra' && policy.name === 'protected-data');
+    case 'GCP':
+      return (workspace as GoogleWorkspace).workspace.bucketName.startsWith('fc-secure');
```
Comment (pshapiro4broad) [inline src/import-data/protected-data-utils.ts]: I like `switch` but I see that if you had instead used `isAzureWorkspace` and `isGoogleWorkspace` then typescript would know what subtype `workspace` is and you wouldn't need to use `as GoolgeWorkspace`. Not sure if there's a good way to do this with `switch`; maybe there's some way to use `workspace.workspace.platform` as a discriminator?

Also, maybe this should be in `workspace-utils`, since it relies on knowledge about the internals of workspaces.

File: src/import-data/ImportDataDestination.ts (hunk 1 above)
Comment (pshapiro4broad) [reply src/import-data/ImportDataDestination.ts:119]: On a similar line, `useWorkspace` hasn't yet been converted to typescript. When it is, will we be able to to type the return value `workspaces` so typescript knows which fields must be present or not based on the arguments? Depending on how many different common cases there are, we may want to create different APIs for each one, vs leaving the API open ended.

(I took a quick look and only saw one other case where the default fields weren't being used, so creating APIs for each custom use wouldn't be too bad.)

**Review by pshapiro4broad — APPROVED** [review]
Looks good

File: src/import-data/ImportDataDestination.ts (hunk 1 above)
Comment (nawatts) [reply src/import-data/ImportDataDestination.ts:119]: Yeah, admittedly here I'm continuing to punt on what is a problem festering across the codebase: code is typed as if we always have the full Workspace object, but we almost always have some partial subset of it.

#4312 converts `useWorkspaces` to TypeScript, but I don't think it addresses the different return types.

File: src/import-data/protected-data-utils.ts (hunk 4 above)
Comment (nawatts) [reply src/import-data/protected-data-utils.ts]: 🤔 Yeah, I think we could use `workspace.workspace.cloudPlatform`.

Some annoyance here caused by _every_ service using different enums for Azure and Google workspaces: `AZURE` and `GCP` for IA code, `Azure` and `Gcp` for workspaces, `azure` and `gcp` for TDR.

File: src/import-data/protected-data-utils.test.ts (hunk 3 above)
Comment (nawatts) [reply src/import-data/protected-data-utils.test.ts:97]: There is a test case that asserts that the default Azure workspace fixture (which has no policies) is not considered protected.

File: src/import-data/protected-data-utils.ts (hunk 4 above)
Comment (nawatts) [reply src/import-data/protected-data-utils.ts]: Done in https://github.com/DataBiosphere/terra-ui/pull/4327/commits/499defd0ae37a4d05d5fcdf88fe93abc6e4596b1.

File: src/import-data/protected-data-utils.test.ts (hunk 3 above)
Comment (nawatts) [reply src/import-data/protected-data-utils.test.ts:97]: Added a test case for a different policy in https://github.com/DataBiosphere/terra-ui/pull/4327/commits/2098e3598bf35f457e91cab76c3ebf32daf0705c.

File: src/import-data/protected-data-utils.ts [hunk 5]
```diff
@@ -68,12 +68,27 @@ export const isProtectedSource = (importRequest: ImportRequest): boolean => {
...
-// For now this also means only GCP workspaces are included.
+/**
+ * Determine whether a workspace is considred protected.
+ *
+ * For Azure workspaces, this checks for the "protected-data" policy.
+ * For Google workspaces, this checks for has enhanced logging - either directly or from an auth domain.
+ *
+ * @param workspace - The workspace.
+ */
 export const isProtectedWorkspace = (workspace: WorkspaceWrapper): boolean => {
```
Comment (cahrens) [inline src/import-data/protected-data-utils.ts:79]: @nawatts did you see that there is also a method to determine this in `workspace-utils`?

```
export const hasProtectedData = (workspace: BaseWorkspace): boolean => containsProtectedDataPolicy(workspace.policies);

export const containsProtectedDataPolicy = (policies: WorkspacePolicy[] | undefined): boolean =>
  _.any((policy) => policy.namespace === 'terra' && policy.name === 'protected-data', policies);

Comment (cahrens) [reply src/import-data/protected-data-utils.ts:79]: It seems like you could at least refactor yours to call `containsProtectedDataPolicy` so that we don't duplicate the logic of the specific policy name.

Comment (nawatts) [reply src/import-data/protected-data-utils.ts:79]: No, I missed that. I can update this function to use that.

Comment (nawatts) [reply src/import-data/protected-data-utils.ts:79]: Done in #4344.
